### PR TITLE
Add crd package

### PR DESCRIPTION
**Pull request by tomahawk28**
Toward https://github.com/giantswarm/roadmap/issues/219

### CODE REVIEW

File: pkg/crd/crd.go [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+		apiextensionsReference: config.ApiextensionsReference,
+		githubClient:           githubClient,
+		provider:               config.Provider,
+	}
+
+	return crdGetter, nil
+}
+
+func (g CRDGetter) LoadCRD(ctx context.Context, group, kind string) (*apiextensionsv1.CustomResourceDefinition, error) {
```
Comment (tomahawk28) [inline pkg/crd/crd.go:80]: This is the function to replace `LoadV1` from `crd` package in apiextensions.

**Comment by tomahawk28** [comment]
Tested in this PR; https://github.com/giantswarm/app-operator/pull/791

**Comment by rossf7** [comment]
Thanks for your efforts here, it's a difficult problem but using GitHub auth for this worries me a lot.

It's ok in a CLI tool or even with Argo but since we have an instance of app-operator per cluster, I think there will be problems even with caching.

I think we should use GitHub Pages for this. We only need to push https://github.com/giantswarm/apiextensions/tree/master/helm/crds-common to the default catalog. In the chartcrd resource we could use helmclient.PullChartTarball and in the app library we just take the path to the tarball.

We'd need to extract the tarball but otherwise, we can use the same logic to get the chart CRD. WDYT?

**Comment by tomahawk28** [comment]
> I think we should use GitHub Pages for this. We only need to push https://github.com/giantswarm/apiextensions/tree/master/helm/crds-common to the default catalog.

If we can use GitHub pages for apiextensions, I think we can access CRD objects by URL like `giantswarm.github.io/apiextensions/tree/master/crds-common` so no need to involve helm chart compressed files. 
But reading the rest of your comment, I think you suggest using the chart compressed file, doesn't it?

> it's a difficult problem but using GitHub auth

It only uses the GitHub auth when the package receives the GitHub token from users so auth is not required to get the CRD, actullay. By the way, CAPI upstream uses a similar approach to download resources from Github's release and we don't need to put GitHub auth there too. 

Last week we had a talk with @tfussell around this topic; https://gigantic.slack.com/archives/C76JX6YLQ/p1622121143032500

In short, I would suggest using a similar approach as CAPI since they already use GitHub client, WDYT?

**Comment by rossf7** [comment]
> If we can use GitHub pages for apiextensions, I think we can access CRD objects by URL like giantswarm.github.io/apiextensions/tree/master/crds-common 

I'd prefer to use this than GH auth but since its a helm chart why can't we push it to a catalog that already has GH pages enabled?

> In short, I would suggest using a similar approach as CAPI since they already use GitHub client

We run app-operator in lots of places including automated tests and long term we want OSS usage to be possible. This introduces a new failure condition we didn't have before. GitHub Pages breaks way less often and we don't have problems pulling chart tarballs from there.

**Comment by tomahawk28** [comment]
>I'd prefer to use this than GH auth but since its a helm chart why can't we push it to a catalog that already has GH pages enabled?

This involves download a chart, unzip it, and finding a file across all directories. Surely we can do this, but if we move our all the resource into GitHub release as CAPI soon or later, all this effort is going to be ineffective. In app-operator, I'm going to download this helm chart file only once in 2~3 hours by using cache, that would not have many troubles for sure. 
WDYT? I mean, we can do as you suggest, but that would take too long to achieve what we first planned.

**Comment by rossf7** [comment]
Yes I guess with plenty of caching in app-operator we can live with this.

**Review by rossf7 — APPROVED** [review]
